datafusion/physical-expr/src/datetime_expressions.rs

```diff
@@ -377,10 +378,10 @@ pub fn date_bin(args: &[ColumnarValue]) -> Result<ColumnarValue> {
 macro_rules! extract_date_part {
     ($ARRAY: expr, $FN:expr) => {
         match $ARRAY.data_type() {
-            DataType::Date32 => {
-                let array = $ARRAY.as_any().downcast_ref::<Date32Array>().unwrap();
-                Ok($FN(array)?)
-            }
+            DataType::Date32 => match as_date32_array($ARRAY) {
+                Ok(array) => Ok($FN(array)?),
+                Err(e) => Err(e),
+            },
```

Comment:
You might be able to make this even more concise like

```suggestion
            DataType::Date32 => $FN(match as_date32_array($ARRAY)?),
```

Reply:
I couldn't find an appropriate way to implement this. It seems that it has to be wrapped around `Ok` or `Err` to return `Result`.